spmd/compiler/fusion.py

```diff
@@ -162,22 +214,448 @@ def _scan_graph_for_fusion_elements(
+
+    def copy_to_buffer(
+        buffer: torch.Tensor, tensor_list: List[torch.Tensor]
+    ) -> torch.Tensor:
+        offset = 0
+        for t in tensor_list:
+            size = t.numel()
+            buffer[offset : offset + size] = t.view(-1)
+            offset += size
+        return buffer
+
+    # setup dummy vars
+    buffer = None
+    if gi.tracing_buffer is None:
+        buffer = torch.empty(buffer_size)
+        gi.tracing_buffer = buffer
+    elif gi.tracing_buffer:
+        buffer = buffer = torch.empty(buffer_size)
+
+    tlist = []
+    for item in copy_list:
+        a = torch.zeros((item.shape[0], item.shape[1]))  # type: ignore
+        tlist.append(a)
+
+    load_gm = make_fx(copy_to_buffer)(buffer, tlist)
+
+    subnodemap = create_graph_node_map(load_gm)
+
+    # update load loop to use main graph items
+    fn_list = []
+    pl_list = []
+    for node in load_gm.graph.nodes:
+        if node.op == OP.PLACEHOLDER:
+            pl_list.append(node)
+        elif node.op == OP.CALL_FUNCTION:
+            fn_list.append(node)
+
+    # create placeholder remapping
+    pl_map: Dict[fx.Node, fx.Node] = {}
+    pl_map[pl_list[0]] = gi.global_buffer_node  # type: ignore
+
+    for i in range(num_fusion_elements):
+        pl_map[pl_list[i + 1]] = in_fe_list[i].grad_tensor_node  # type: ignore
+
+    insert_node = in_fe_list[-1].comm_node
+
+    def remap_copy_args(in_node: fx.Node) -> fx.Node:
+        out_node = in_node
+        if in_node in pl_map:
+            out_node = pl_map[in_node]  # type: ignore
+        elif in_node in value_remap:
+            out_node = value_remap[in_node]
+        return out_node
+
+    value_remap = {}
+    with gm.graph.inserting_before(insert_node):
+        for innernode in load_gm.graph.nodes:
+            if innernode.op in [OP.PLACEHOLDER, OP.OUTPUT]:
+                continue
+            value_remap[innernode] = gm.graph.node_copy(
+                innernode, remap_copy_args
+            )
+
+    _update_new_copy_nodes_users(value_remap)
+
+    # update allreduce to use buffer
+    # (we currently don't) have to make our own all_reduce/comm_wait section
+    # # TODO - pg group matching
+    # _build_buffer_comm_graph(gm, gi)
+
+    buffer_comm_node = in_fe_list[-1].comm_node
+    buffer_comm_node.update_arg(0, [buffer_node])  # type: ignore
+
+
+def _build_buffer_comm_graph(
+    gi: GraphInfo, gm: fx.GraphModule
+) -> fx.GraphModule:
+    """have to make our own all_reduce and wait subgraph for buffer"""
+    # from torch.distributed._spmd.comm_tensor import CommTensor
+    from torch.distributed.distributed_c10d import (  # ProcessGroup,; Work,; all_reduce,
+        ReduceOp,
+        _get_default_group,
+    )
+
+    buffer_size = gi.global_buffer_size
+
+    def dummy_add(
+        grad_buffer: torch.Tensor, zero: torch.Tensor
+    ) -> torch.Tensor:
+        return grad_buffer + zero
+
+    grad_buffer: torch.Tensor = torch.empty(buffer_size)
```

Comment:
i would probably also have just created an fx.Graph, added 2 placeholders to it with names, and then inserted a call_function(torch.add, a, b) and an output node.  Not necessary to allocate real tensors and trace a graph.  possibly also fuse this whole function inline into wherever it is called, so you don't have to iterate this new graph and copy its nodes into another graph- just create them in the right place up front?

Reply:
sorry for the confusion here - this is a half completed function that atm is not used.  We do a full trace b/c ultimately this has to be wrapped with a CommTensor to produce and wait on the comm work handle.  
For now I am re-using existing wait comms in the graph rather than writing our own which was why this function is only half complete. 
I've updated in the comments that this is a stub to make it clearer.